rpcs3/util/expr.hpp

```diff
@@ -0,0 +1,281 @@
+struct utf32_cmd<U'?', S, I>
+{
+	template <char32_t Cmd, usz J, usz L = 0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) skip(u64 ctrs[], SS&&... stack)
+	{
+		if constexpr (Cmd == U'\\' && L == 0)
+		{
+			return utf32_cmd<S.data[J], S, J + 1>::eval(ctrs, std::forward<SS>(stack)...);
+		}
+
+		if constexpr (Cmd == U'@')
+		{
+			return skip<S.data[J], J + 1, L + 1>(ctrs, std::forward<SS>(stack)...);
+		}
+
+		if constexpr (Cmd == U';')
+		{
+			if constexpr (L > 0)
+			{
+				return skip<S.data[J], J + 1, L - 1>(ctrs, std::forward<SS>(stack)...);
+			}
+			else
+			{
+				return;
+			}
+		}
+
+		skip<S.data[J], J + 1, L>(ctrs, std::forward<SS>(stack)...);
+	}
+
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		if (s0)
+		{
+			utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<SS>(stack)...);
+		}
+		else
+		{
+			skip<S.data[I], S, I + 1>(std::forward<SS>(stack)...);
+		}
+	}
+};
+
+// Duplicate top stack value
+template <utf32_expr S, usz I>
+struct utf32_cmd<U':', S, I>
+{
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<S0>(s0), std::forward<S0>(s0), std::forward<SS>(stack)...);
+	}
+};
+
+// Duplicate two top stack values
+template <utf32_expr S, usz I>
+struct utf32_cmd<U'#', S, I>
+{
+	template <typename S0, typename S1, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, S1&& s1, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<S0>(s0), std::forward<S0>(s1), std::forward<S0>(s0), std::forward<S1>(s1), std::forward<SS>(stack)...);
+	}
+};
+
+// Swap two top stack values
+template <utf32_expr S, usz I>
+struct utf32_cmd<U's', S, I>
+{
+	template <typename S0, typename S1, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, S1&& s1, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<S1>(s1), std::forward<S0>(s0), std::forward<SS>(stack)...);
+	}
+};
+
+// Rotate stack so the top becomes the bottom
+template <utf32_expr S, usz I>
+struct utf32_cmd<U',', S, I>
+{
+	template <typename S0, typename... SS>
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, SS&&... stack)
+	{
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, std::forward<SS>(stack)..., std::forward<S0>(s0));
+	}
+};
+
+#define REG_BIN_OP(ch, _func)\
+template <utf32_expr S, usz I>\
+struct utf32_cmd<ch, S, I>\
+{\
+	template <typename S0, typename S1, typename... SS>\
+	static FORCE_INLINE constexpr SAFE_BUFFERS(auto) eval(u64 ctrs[], S0&& s0, S1&& s1, SS&&... stack)\
+	{\
+		return utf32_cmd<S.data[I], S, I + 1>::eval(ctrs, (_func)(std::forward<S1>(s1), std::forward<S0>(s0)), std::forward<SS>(stack)...);\
```

Comment:
Why are you swapping s1 and s0 here? Just trying to get my head around this stuff, it is interesting...

Reply:
s0 is stack head, and IIRC "natural" way of writing the stack is bottom-to-top, but specifics of C++ to decompose parameter pack (which stores the stack) makes it much easier to have it top-to-bottom.

Reply:
On the other hand, it can be done as s0*s1 instead, since args are pushed in this order.

Reply:
I don't think i quite understood you. Bottom-to-top or top-to-bottom is underlying iml, and is irrelevant as long as stack behaves like it should. Arguments intro arguments pack are pushed from last to first. At the start you have (x <- stack head, y), the current metastring substitutes them for their elements and swap them 2 times so (x[0] <-stack head, y[0]) but you swap x[0] and y[0] which in the end results in front ordering compared to the original comparator. 

Original comparator sorts from least to greatest with respect to element 0, the metastring one does the opposite, I checked on a simple tuple vector.

Reply:
Also "s0s0<" could be done like "0s0>"

Reply:
No, it's supposed to be <. And the order after two swaps is correct. I incorrectly swapped s1 and s0 in the template.

Reply:
i mean in "0s0>" if you swap only once so then before the `>` op the stack is (y[0], x[0]) so you softer correcting `REG_BIN_OP` that results in `y[0] > x[0]`. and "s0s0<" (with correct `REG_BIN_OP`) results in `x[0] < y[0]` which is the same thing. I know it is irrelevant since meta->op transformation is done by compiler.

Reply:
btw why all the `if constexpr` inside `constexpr` functions? Is it not redundant since all expressions are evaluated at compile-time anyway?